dataRetriever/resolvers/miniblockResolver.go

```diff
@@ -0,0 +1,202 @@
+		return nil, dataRetriever.ErrNilDataPacker
+	}
+
+	mbResolver := &miniblockResolver{
+		TopicResolverSender: arg.SenderResolver,
+		miniBlockPool:       arg.MiniBlockPool,
+		miniBlockStorage:    arg.MiniBlockStorage,
+		dataPacker:          arg.DataPacker,
+		messageProcessor: messageProcessor{
+			marshalizer:      arg.Marshalizer,
+			antifloodHandler: arg.AntifloodHandler,
+			topic:            arg.SenderResolver.RequestTopic(),
+			throttler:        arg.Throttler,
+		},
+	}
+
+	return mbResolver, nil
+}
+
+// ProcessReceivedMessage will be the callback func from the p2p.Messenger and will be called each time a new message was received
+// (for the topic this validator was registered to, usually a request topic)
+func (mbRes *miniblockResolver) ProcessReceivedMessage(message p2p.MessageP2P, fromConnectedPeer p2p.PeerID) error {
+	err := mbRes.canProcessMessage(message, fromConnectedPeer)
+	if err != nil {
+		return err
+	}
+
+	mbRes.throttler.StartProcessing()
+	defer mbRes.throttler.EndProcessing()
+
+	rd, err := mbRes.parseReceivedMessage(message)
+	if err != nil {
+		return err
+	}
+
+	switch rd.Type {
+	case dataRetriever.HashType:
+		var buff []byte
+		buff, err = mbRes.resolveMbRequestByHash(rd.Value)
+		if err != nil {
+			return err
+		}
+		return mbRes.Send(buff, message.Peer())
+	case dataRetriever.HashArrayType:
+		return mbRes.resolveMbRequestByHashArray(rd.Value, message.Peer())
+	default:
+		return dataRetriever.ErrRequestTypeNotImplemented
+	}
+}
+
+func (mbRes *miniblockResolver) resolveMbRequestByHash(hash []byte) ([]byte, error) {
+	mb, err := mbRes.fetchMbAsByteSlice(hash)
+	if err != nil {
+		return nil, err
+	}
+
+	buffToSend, err := mbRes.marshalizer.Marshal(&batch.Batch{Data: [][]byte{mb}})
+	if err != nil {
+		return nil, err
+	}
+
+	return buffToSend, nil
+}
+
+func (mbRes *miniblockResolver) fetchMbAsByteSlice(hash []byte) ([]byte, error) {
+	value, ok := mbRes.miniBlockPool.Peek(hash)
+	if ok {
+		txBuff, err := mbRes.marshalizer.Marshal(value)
+		if err != nil {
+			return nil, err
+		}
+		return txBuff, nil
+	}
+
+	return mbRes.miniBlockStorage.SearchFirst(hash)
+}
+
+func (mbRes *miniblockResolver) resolveMbRequestByHashArray(mbBuff []byte, pid p2p.PeerID) error {
+	b := batch.Batch{}
+	err := mbRes.marshalizer.Unmarshal(&b, mbBuff)
+	if err != nil {
+		return err
+	}
+	hashes := b.Data
+
+	mbsBuffSlice := make([][]byte, 0, len(hashes))
+	for _, hash := range hashes {
+		mb, errFetch := mbRes.fetchMbAsByteSlice(hash)
+		if errFetch != nil {
+			err = errFetch
+			log.Trace("fetchTxAsByteSlice missing",
+				"error", err.Error(),
+				"hash", hash)
+			continue
+		}
+		mbsBuffSlice = append(mbsBuffSlice, mb)
+	}
+
+	buffsToSend, errPack := mbRes.dataPacker.PackDataInChunks(mbsBuffSlice, maxBuffToSendBulkTransactions)
+	if errPack != nil {
+		return errPack
+	}
+
+	for _, buff := range buffsToSend {
+		errSend := mbRes.Send(buff, pid)
+		if errSend != nil {
+			return errSend
+		}
+	}
+
+	return err
```

Comment:
so the only way that the returned err isn't nil is when you can't fetch at least one miniblock but you managed to send  the rest of them. is this ok ? Maybe implement a counter for how many miniblocks you managed to fetch&send and return it also. The way it is now is not intuitive

Reply:
Yes, this is the intended behavior. Renamed variables for improved code understanding. Added errors found